### PR TITLE
899 - Datagrid Xss, Nested Data and Formatters and Virtual Scroll

### LINKED ISSUE #627: IdsVirtualScroll: Remaining Bugs
**Describe the bug**
This describes the remaining things to check in virtual scroll

**To Reproduce**

- [x] Navigate to: https://main.wc.design.infor.com/ids-data-grid/virtual-scroll.html
- [x] List should resize to the bottom minus padding
- [x] Aria attributes are undefined
- [x]  Keyboard - arrow doesn't work
- [x] click event delegation?
- [x] cant scroll horizontal if many columns
- [x]   row number formatter not working `rowNumber`

### LINKED ISSUE #862: IdsDataGrid: Add ability to display nested data
**Is your feature request related to a problem or use case? Please describe.**
Currently it is not possible to push nested data into the `ids-data-grid` columns. 

**Describe the solution you'd like**
The solution should work as follows:

Example data shape:
```
{ 
    path: 'ids-alert', 
    loadChildren: () => import('./components/ids-alert/ids-alert.module').then(m => m.IdsAlertModule),
    data: {
      title: 'IDS Alert',
      category: 'Form Inputs',
      description: 'Alert Indicator/icon'
    },
  },
```

Implementation
```
this.columns.push({
      id: 'type',
      name: 'Type',
      field: 'data.title',
      sortable: true,
      formatter: this.table.nativeElement.formatters.text
});
```

**Describe alternatives you've considered**
n/a

**Additional context**
n/a

### LINKED ISSUE #899: IdsDataGrid: XSS issue in cells
**Describe the bug**
Review of IdsDataGrid usage with the M3 team revealed that it's currently possible to inject script tags into IdsDataGrid cells.  They were doing this by loading existing data into a cell (not necessarily from editing a cell).

**How to Reproduce**
We can't currently reproduce in the WC project -- we need to create some test cases for these conditions (may be able to [re-use the ones from Enterprise](https://main-enterprise.demo.design.infor.com/components/datagrid/test-xss-prevention.html))

- Create an example page containing an IDS Datagrid with a pre-determined dataset.
- In this dataset, add cell content with JS script tags, like one that creates an alert with arbitrary text.
- Open the new example page. When the data grid is invoked, an alert will be displayed.

**Additional Context**:
- Discovered by the M3 Team (cc @inforrpunay)
- Should be completed along with #737

### PR DESCRIPTION
**Pull request by tmcconechy**
**Explain the details for making this change. What existing problem does the pull request solve?**
This PR adds a few general features needed as a base and starting point for #737. The following features and fixes are added.

- Ability to supported nested data in the dataset `data.child.setting` for example
- Full test covered (enabled paging tests)
- Xss prevention
- Checkbox Formatter
- Custom Formatter
- Fixes to Virtualization to support selection and selection indexes
- Refactor htmlEntities to escapteHTML (double functions existed)

**Related github/jira issue (required)**:
Fixes #899 
Fixes #862 
Fixes #627
Some of #737 

**Steps necessary to review your pull request (required)**:
- go to http://localhost:4300/ids-data-grid/data-nested.html and see that all columns are rendering a value even on tho some data is nested 2 and 3 levels deep in the data array
- go to http://localhost:4300/ids-data-grid/xss-prevention.html and check the console for no errors (should be no csp errors). This page will now render with "xss" "attacks" in it escaping the data
- tests and checks all pass (on the ci)
- go to http://localhost:4300/ids-data-grid/columns-formatters.html and observe the new checkbox formatter and the custom formatter (last one). The custom formatter shows how to roll your own formatter.
- go to http://localhost:4300/ids-data-grid/virtual-scroll.html
- select a row near the top and bottom
- check the console that the row's selectedRows index is correct
- scroll back up and see the selection is maintained
- try and sort the columns

**Included in this Pull Request**:
- [x] An e2e or functional test for the bug or feature.
- [x] A note to the change log.

### CODE REVIEW

**Review by jmacaluso711 — DISMISSED** [review]
Looks good 👍

**Comment by EdwardCoyle** [comment]
Also linking this to #900, since this partially addresses that issue

File: src/components/ids-virtual-scroll/ids-virtual-scroll.ts [hunk 1]
```diff
@@ -23,7 +22,8 @@ export default class IdsVirtualScroll extends Base {
   }
 
   connectedCallback() {
-    this.datasource = new IdsDataSource();
+    // Array is a pointer to a datasource in a parent component
+    this.datasource = [];
```
[the remark this answers is not in the document]
Comment (tmcconechy) [reply src/components/ids-virtual-scroll/ids-virtual-scroll.ts]: Yeah the problem was that the IdsDatasource clones the array. So in this case i want it to remain a "link" to the datasource/array on the parent component. But i see the point so i made this into a type on the root instead. i think thats what you mean? here: https://github.com/infor-design/enterprise-wc/pull/934/commits/7ecd597e3167ae55de5847a175b1237320b3981b

The bug it fixes by not recloning is 
- go to http://localhost:4300/ids-data-grid/virtual-scroll.html
- select first one and scroll to the very bottom and select one at the bottom, then scroll back up (all selections should be retained)

**Review by deep7102 — APPROVED** [review]
LGTM 🍌

**Comment by tmcconechy** [comment]
Merging as approved previously by 2/3

**Pull request merged**